tests/e2e/json_utils.go

```diff
@@ -0,0 +1,398 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// stores a proposal as a raw json, together with its type
+type ProposalAndType struct {
+	RawProposal json.RawMessage
+	Type        string
+}
+
+// duplicated from the ChainState with a minor change to the Proposals field
+type ChainStateWithProposalTypes struct {
+	ValBalances                    *map[ValidatorID]uint
+	ValPowers                      *map[ValidatorID]uint
+	RepresentativePowers           *map[ValidatorID]uint
+	Params                         *[]Param
+	Rewards                        *Rewards
+	ConsumerChains                 *map[ChainID]bool
+	AssignedKeys                   *map[ValidatorID]string
+	ProviderKeys                   *map[ValidatorID]string
+	ConsumerChainQueueSizes        *map[ChainID]uint
+	GlobalSlashQueueSize           *uint
+	RegisteredConsumerRewardDenoms *[]string
+	Proposals                      *map[uint]ProposalAndType // the only thing changed from the real ChainState
```

Comment:
Field shadowing might be a solution to have a `ChainState` while changing the Proposal field type, see https://stackoverflow.com/questions/56982842/how-to-selectively-marshal-json-for-a-struct

Reply:
Thanks, amazing. This took a bit of effort but I managed to do it with shadowing. It needs a hack, since I need to avoid calling the same custom MarshalJSON method, so I have to define a copy of the ChainState struct, but I can just type alias it. See [f8c8329](https://github.com/cosmos/interchain-security/pull/1314/commits/f8c8329d4cf00040c965b4281b01250880ffa7fe)

Reply:
Okay, got it. I knew it'd require a bit of hacking, but it's much cleaner now!
Solid work!